service/udp_linux.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2018 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package service
+
+import (
+	"context"
+	"net"
+	"syscall"
+	"unsafe"
+
+	onet "github.com/Jigsaw-Code/outline-ss-server/net"
+	"golang.org/x/sys/unix"
+)
+
+func ListenUDP(network string, laddr *net.UDPAddr) (onet.UDPPacketConn, error) {
+	lc := &net.ListenConfig{
+		Control: func(network, address string, c syscall.RawConn) error {
+			return c.Control(func(fd uintptr) {
+				// Always set IP_PKTINFO
+				if err := unix.SetsockoptInt(int(fd), unix.IPPROTO_IP, unix.IP_PKTINFO, 1); err != nil {
```

Comment:
Won't this fail for IPv6?

Reply:
It won't, because Go defaults to dual-stack IPv6 sockets. This is also the case with shadowbox, where this function is only called once with network `udp6`. If you don't set `IP_PKTINFO` on dual-stack IPv6 sockets, you won't receive the necessary OOB data when an IPv4 client connects.